evm_loader/program/src/evm/database.rs

```diff
@@ -51,3 +49,244 @@ pub trait Database {
         is_static: bool,
     ) -> Option<Result<Vec<u8>>>;
 }
+
+/// Provides convenience methods that can be implemented in terms of `Database`.
+#[maybe_async(?Send)]
+pub trait DatabaseExt {
+    /// Returns whether an account exists, which is also referred to as the account being non-empty,
+    /// in accordance with https://eips.ethereum.org/EIPS/eip-161.
+    async fn account_exists(&self, address: Address, chain_id: u64) -> Result<bool>;
+
+    /// Returns the code has for an address in accordance with https://eips.ethereum.org/EIPS/eip-1052.
+    async fn code_hash(&self, address: Address, chain_id: u64) -> Result<[u8; 32]>;
+}
+
+#[maybe_async(?Send)]
+impl<T: Database> DatabaseExt for T {
+    async fn account_exists(&self, address: Address, chain_id: u64) -> Result<bool> {
+        Ok(self.nonce(address, chain_id).await? > 0 || self.balance(address, chain_id).await? > 0)
+    }
+
+    async fn code_hash(&self, address: Address, chain_id: u64) -> Result<[u8; 32]> {
+        // The function `Database::code` returns a zero-length buffer if the account exists with
+        // zero-length code, but also when the account does not exist. This makes it necessary to
+        // also check if the account exists when the returned buffer is empty.
+        //
+        // We could simplify the implementation by checking if the account exists first, but that
+        // would lead to more computation in what we think is the common case where the account
+        // exists and contains code.
+        let code = self.code(address).await?;
+        let bytes_to_hash: Option<&[u8]> = if !code.is_empty() {
+            Some(&*code)
+        } else if self.account_exists(address, chain_id).await? {
+            Some(&[])
+        } else {
+            None
+        };
+
+        Ok(bytes_to_hash.map_or([0; 32], |bytes| {
+            solana_program::keccak::hash(bytes).to_bytes()
+        }))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+
+    use super::*;
+
+    struct TestDatabaseEntry {
+        balance: U256,
+        nonce: u64,
+        code: Vec<u8>,
+    }
+
+    #[allow(dead_code)]
+    struct TestDatabase(HashMap<Address, TestDatabaseEntry>);
+
+    const NON_EXISTENT_ADDRESS: Address = Address([0xaa; 20]);
+    const EMPTY_CODE_ADDRESS: Address = Address([0xbb; 20]);
+    const NON_EMPTY_CODE_ADDRESS: Address = Address([0xcc; 20]);
+
+    impl Default for TestDatabase {
+        fn default() -> Self {
```

Comment:
I suggest just creating a `new` inherent method instead of implementing `Default`, since `Default` for `TestDatabase` actually should contain an empty map.

Reply:
I think default is appropriate given the type name is `TestDatabase`. 

The `Default` docs say: 

> A trait for giving a type a useful default value.